be/src/runtime/load_channel_mgr.cpp

```diff
@@ -75,94 +72,52 @@ Status LoadChannelMgr::init(MemTracker* mem_tracker) {
 
-            channel.reset(new LoadChannel(load_id, job_timeout_s, std::move(job_mem_tracker)));
+            channel.reset(new LoadChannel(this, load_id, job_timeout_s, std::move(job_mem_tracker)));
             _load_channels.insert({load_id, channel});
         }
     }
-
-    RETURN_IF_ERROR(channel->open(params));
-    return Status::OK();
+    channel->open(cntl, request, response, done_guard.release());
 }
 
-static void dummy_deleter(const CacheKey& key, void* value) {}
-
-Status LoadChannelMgr::add_chunk(const PTabletWriterAddChunkRequest& request,
-                                 google::protobuf::RepeatedPtrField<PTabletInfo>* tablet_vec,
-                                 int64_t* wait_lock_time_ns) {
+void LoadChannelMgr::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChunkRequest& request,
+                               PTabletWriterAddBatchResult* response, google::protobuf::Closure* done) {
+    VLOG(2) << "Current memory usage=" << _mem_tracker->consumption() << " limit=" << _mem_tracker->limit();
+    ClosureGuard done_guard(done);
     UniqueId load_id(request.id());
-    // 1. get load channel
-    std::shared_ptr<LoadChannel> channel;
-    {
-        std::lock_guard<std::mutex> l(_lock);
-        auto it = _load_channels.find(load_id);
-        if (it == _load_channels.end()) {
-            auto* handle = _lastest_success_channel->lookup(load_id.to_string());
-            // success only when eos be true
-            if (handle != nullptr) {
-                _lastest_success_channel->release(handle);
-                if (request.has_eos() && request.eos()) {
-                    return Status::OK();
-                }
-            }
-            return Status::InternalError(
-                    strings::Substitute("fail to add batch in load channel. unknown load_id=$0", load_id.to_string()));
-        }
-        channel = it->second;
-    }
-
-    // 3. add batch to load channel
-    // batch may not exist in request(eg: eos request without batch),
-    // this case will be handled in load channel's add batch method.
-    RETURN_IF_ERROR(channel->add_chunk(request, tablet_vec));
-
-    // 4. handle finish
-    if (channel->is_finished()) {
-        LOG(INFO) << "Removing finished load channel load id=" << load_id;
-        {
-            std::lock_guard<std::mutex> l(_lock);
-            _load_channels.erase(load_id);
-            auto* handle = _lastest_success_channel->insert(load_id.to_string(), nullptr, 1, dummy_deleter);
-            _lastest_success_channel->release(handle);
-        }
-        VLOG(1) << "Removed load channel load id=" << load_id;
+    auto channel = _find_load_channel(load_id);
+    if (channel != nullptr) {
+        channel->add_chunk(cntl, request, response, done_guard.release());
+    } else {
+        response->mutable_status()->set_status_code(TStatusCode::INTERNAL_ERROR);
+        response->mutable_status()->add_error_msgs("no associated load channel");
```

Comment:
better to add load_id or something that can stand for this load job to error message

Reply:
It's the caller's duty